### PR TITLE
dbapi: Add db.collection.name, use connection kwargs for connection attributes

### PR DESCRIPTION
**Pull request by keithZmudzinski**
# Description

This pull request adds 3 features to the opentelemetry-instrumentation-dbapi library.
1. Includes `db.collection.name` in the span name if possible. This is in line with the semantic conventions for database span names: https://opentelemetry.io/docs/specs/semconv/database/database-spans/#name
2. Adds the `db.collection.name` attribute to the span if possible.
3. Uses more sources for the `db.user`, `net.peer.name`, and `net.peer.port` span attributes.
    - I found that when using opentelemetry-instrumentation-dbapi to instrument `MySQLdb`, there wasn't a way to get the host from the instantiated Connection class. However, the host is present in the kwargs used to create the connection. If we parse the kwargs for these attributes, we are more likely to get them.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

I've used these changes in a local project using the console exporter and New Relic across a variety of sql connections. Confirmed the expected attributes were populated.
Added unit tests to verify the changes.

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated (NA)

### CODE REVIEW

**Comment by keithZmudzinski** [comment]
I'm working on getting the CLA signed, but for what it's worth I've passed dbapi tests and linting locally

**Comment by keithZmudzinski** [comment]
Making progress on the CLA, I've got a person who said they'd be the CLA manager

**Comment by keithZmudzinski** [comment]
We've got a CLA manager, just awaiting approval now

**Comment by keithZmudzinski** [comment]
Sweet, I just got approved. Good to review now I believe

**Comment by keithZmudzinski** [comment]
@emdneto Do you know who I could ping to get a review on this?

**Review by emdneto — COMMENTED** [review]
Thanks for the PR.

Since the semantic conventions for Database Calls are Experimental, I understand we shouldn't change the version of the database conventions that they emit until they are marked stable. This implies not adding new attributes that only exist in the new versions of the semantic conventions. 

However, I'm ok with adding the DB name to the span name because in [v1.11.0](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.11.0/specification/trace/semantic_conventions/database.md) it is allowed. 

You may want to consider addressing the linting issues and setting up pre-commit in your environment.

**Comment by keithZmudzinski** [comment]
@emdneto Linting and test failures addressed:

I ran the `dbapi` tests for `py10,11,12`, all passed:
image
image
image


I've run `isort` and `black` and the dbapi files:
image

**Comment by keithZmudzinski** [comment]
I'm having some trouble running the `docker-tests` on my machine. Before I dive into it @emdneto , I just wanted to make sure that the [failing test](https://github.com/zyBooks/opentelemetry-python-contrib/actions/runs/11336288525/job/31525959858#step:5:436) isn't flaky as far as you know?

File: tests/opentelemetry-docker-tests/tests/pymysql/test_pymysql_functional.py [hunk 1]
```diff
@@ -84,23 +84,23 @@ def test_execute(self):
         stmt = "CREATE TABLE IF NOT EXISTS test (id INT)"
         with self._tracer.start_as_current_span("rootSpan"):
             self._cursor.execute(stmt)
-        self.validate_spans("CREATE")
+        self.validate_spans("CREATE test")
```
Comment (emdneto) [inline tests/opentelemetry-docker-tests/tests/pymysql/test_pymysql_functional.py:87]: @keithZmudzinski, they are legit tests that need to be fixed if we want this PR to be merged.

File: instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py [hunk 2]
```diff
@@ -373,14 +381,19 @@ def _populate_span(
...
+
         span.set_attribute(
             SpanAttributes.DB_SYSTEM, self._db_api_integration.database_system
         )
         span.set_attribute(
             SpanAttributes.DB_NAME, self._db_api_integration.database
         )
         span.set_attribute(SpanAttributes.DB_STATEMENT, statement)
+        if collection_name:
+            span.set_attribute(DB_COLLECTION_NAME, collection_name)
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py:396]: But as I said in the comments before, I wouldn't introduce new semantic convention attributes in this PR.

File: instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py [hunk 3]
```diff
@@ -284,12 +287,17 @@ def wrapped_connection(
...
         return get_traced_connection_proxy(connection, self)
 
-    def get_connection_attributes(self, connection):
-        # Populate span fields using connection
+    def get_connection_attributes(self, connection, kwargs=None):
+        # Populate span fields using kwargs and connection
         for key, value in self.connection_attributes.items():
+            # First set from kwargs
+            if kwargs and value in kwargs:
+                self.connection_props[key] = kwargs.get(value)
```
Comment (tammy-baylis-swi) [inline instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py:298]: Thanks for putting in this PR!

> 3. Uses more sources for the db.user, net.peer.name, and net.peer.port span attributes.

Could some tests be added to show expectations for this?

File: instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py [hunk 4]
```diff
@@ -391,12 +404,32 @@ def _populate_span(
...
             span.set_attribute("db.statement.parameters", str(args[1]))
 
-    def get_operation_name(self, cursor, args):  # pylint: disable=no-self-use
+    def get_span_name(self, cursor, args):
+        operation_name = self.get_operation_name(cursor, args)
+        statement = self.get_statement(cursor, args)
+        collection_name = CursorTracer.get_collection_name(statement)
+        return " ".join(
+            name for name in (operation_name, collection_name) if name
+        )
```
Comment (tammy-baylis-swi) [inline instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py:413]: Do these changes also implement these parts of the [spec](https://opentelemetry.io/docs/specs/semconv/database/database-spans/#name)?

> If there is no (low-cardinality) db.operation.name available, database span names SHOULD be [{target}](https://opentelemetry.io/docs/specs/semconv/database/database-spans/#target-placeholder).
> 
> If neither {db.operation.name} nor {target} are available, span name SHOULD be {db.system}.

I think Yes for the first point, would be good to see some test cases for that as well.

**Comment by tammy-baylis-swi** [comment]
Hi @keithZmudzinski, we talked about Otel database semantic conventions in today's Python SIG meeting. The changes in PR are appreciated and will be needed at some point. However, because the database semconv is not yet stable (currently it's [release candidate](https://opentelemetry.io/docs/specs/semconv/database/database-spans/)), we will wait before making the proposed updates for the span name and attributes changes. This is a topic we will be bringing up with the Otel TC for guidance.

**Comment by keithZmudzinski** [comment]
@tammy-baylis-swi Thanks for the update, and for the feedback, I really appreciate it.

Should I let this PR stay put, or close it out and open a new one when the database conventions are marked as STABLE?

**Comment by tammy-baylis-swi** [comment]
@keithZmudzinski I'd say it's easiest to Close this one and start a fresh PR later, since this monorepo tends to get a lot of merge conflicts introduced.

**Comment by aowczarek619** [comment]
I think database semconv is already stable https://opentelemetry.io/docs/specs/semconv/db/database-spans/

**Pull request closed without merging**